### PR TITLE
improvement: Change the stamps to timewrapped ones

### PR DESCRIPTION
**Pull request by tgodzik**
Previously we would use last modfied stamps for classfiles (and other products) and hashes for jars. Now, we use hashes for both, but only if the modified time changed.

Should help with https://github.com/scalameta/metals/issues/4623

### CODE REVIEW

**Comment by tgodzik** [comment]
Locally run benchmark seems to produce the same timings and I am running the server one to confirm.

File: backend/src/main/scala/sbt/internal/inc/bloop/internal/BloopStamps.scala [hunk 1]
```diff
@@ -34,8 +34,11 @@ object BloopStamps {
   def directoryHash(path: Path): FileHash = FileHash.of(path, directoryHash)
   def isDirectoryHash(fh: FileHash): Boolean = fh.hash == directoryHash
 
-  def forHash(file: VirtualFileRef): Hash = {
-    fromBloopHashToZincHash(ByteHasher.hashFileContents(converter.toPath(file).toFile()))
+  def forHash(fileRef: VirtualFileRef): Hash = {
+    val file = converter.toPath(fileRef).toFile()
+    if (file.exists())
```
Comment (Duhemm) [inline backend/src/main/scala/sbt/internal/inc/bloop/internal/BloopStamps.scala:39]: Why do we need to check if the file exists now but not before?

**Review by Duhemm — APPROVED** [review]
Looks good to me, thank you!

I'll try that out and let you know how that may affect the performance on larger builds. Don't wait for me though 😄

**Comment by bloopoid** [comment]
Performance test finished successfully.

Benchmarks is based on merging with master

**Comment by bloopoid** [comment]
Performance test finished successfully.

Benchmarks is based on merging with master

**Comment by bloopoid** [comment]
Performance test finished successfully.

Benchmarks is based on merging with master

**Pull request merged**